### PR TITLE
[WIP] Bl/inflight quorum

### PR DESCRIPTION
**Pull request by ben-lerner**
[WIP] - Logic is done and probably works, preliminary feedback would be appreciated. Currently fixing old tests and adding new ones.

Add support for bootstrapping to the write client. While bootstrapping the initializing and leaving shards count as a single shard towards a quorum, and only if the client successfully writes to both of them.

### CODE REVIEW

**Comment by xichen2020** [comment]
Macro comment: Looks like there are both refactoring changes (which are presumably no-op changes as far as functionality is concerned) and functionality changes in this PR. Would be nice if we can separate those into two diffs to make easier for reviewers.

File: client/metadata_iter.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		return false
+	}
+	m, more := <-it.inputCh
+	if !more {
+		it.err = <-it.errCh
+		it.done = true
+		return false
+	}
+	it.host = m.peer.Host()
+	var zeroed block.Metadata
```
Comment (martin-mao) [inline client/metadata_iter.go]: Why not just use nil or move this to const somewhere?

File: client/enqueue_channel.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+func newEnqueueChannel(m *streamFromPeersMetrics) *enqueueChannel {
+	c := &enqueueChannel{
+		peersMetadataCh: make(chan []*blocksMetadata, 4096),
+		closed:          0,
+		metrics:         m,
+	}
+	go func() {
+		for atomic.LoadInt64(&c.closed) == 0 {
+			m.blocksEnqueueChannel.Update(int64(len(c.peersMetadataCh)))
```
Comment (martin-mao) [inline client/enqueue_channel.go]: Is this not atomic.LoadUint64(&c.enqueued) - atomic.LoadUint64(&c.processed)?

File: client/host_queue.go [hunk 3]
```diff
@@ -192,37 +190,36 @@ func (q *queue) drain() {
...
-				currBatchElementsByNamespace[namespace.Hash()] = currBatchElements
+				currWriteOpsByNamespace[hash] = currWriteOps
+				currBatchElementsByNamespace[hash] = currBatchElements
 
 				if len(currWriteOps) == writeBatchSize {
 					// Reached write batch limit, write async and reset
 					q.asyncWrite(wgAll, namespace, currWriteOps, currBatchElements)
-					currWriteOpsByNamespace[namespace.Hash()] = nil
-					currBatchElementsByNamespace[namespace.Hash()] = nil
+					currWriteOpsByNamespace[hash] = nil
```
Comment (kobolog) [inline client/host_queue.go]: This and in other places should probably be delete(currWriteOpsByNamespace, hash), I think?

**Comment by robskillington** [comment]
Hey some of the refactors are good but conflict with changes that have already been merged to master.  Can you reduce this to the minimal change to implement this?  We can do some of the refactors that you've done here as a separate independent change.  

This will help us ship faster and keep this change to a minimal set we can audit and review more carefully.

**Comment by ben-lerner** [comment]
Will do.

**Pull request closed without merging**